encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/FieldRef.java

```diff
@@ -0,0 +1,72 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+import androidx.annotation.NonNull;
+
+//TODO: support required, optional.
+public abstract class FieldRef<T> {
+  @NonNull
+  protected String name;
+
+  @NonNull
+  public String getName() {
+    return name;
+  }
+
+  protected FieldRef(@NonNull String name) {
+    this.name = name;
+  }
+
+  public static class Primitive<T> extends FieldRef<T> {
```

Comment:
Each `ObjectDecoder` keeps its decoding **logics** into a `ObjectDecoderContextIml` where each field name corresponds to a `FieldRef`. The decoding **logic** inside `ObjectDecoderContextIml` is simply an container `Map<String, FieldRef<?>> refs` which refers each field name and its corresponding `FieldRef`.

Each `ObjectDecoder` keeps its decoding **value** from given `inputStream` into a `CreationContext`. The decoding **values** inside `CreationContext` is also an container `Map<FieldRef<?>, Object> ctx` which contains all the values the FieldRef are refering to.

Therefore, whenever we are given the new `InputStream`, We wouldn't create `ObjectDecoderContextIml` again, but only update the values each `FieldRef` pointing to.

The reason we need a `rawType` inside `Primitive<T>` or `ClassToken` inside `Boxed<T>` is that:
each `ObjectDecoderContextIml` should not only knows where each field is referring to but also its anticipated type. Therefore, I nested their type information into `FiedRef`, getting these types information make it possible to determine which method to use when looping the json data, such as `nextInt(), nextBoolean, etc.` 

The reason I make `Primitive` and `Boxed` extennd `FieldRef` is that
no matter they are primitive or object, they are field references. Besides, I want to make the containers inside  `ObjectDecoderContextIml` and `CreationContext` be simple `Map<String, FieldRef<?>>`

@vkryachko 
wdyt? are there any better ways to do it?

Reply:
> The reason we need a rawType inside Primitive<T> or ClassToken inside Boxed<T>

I see that makes sense, I would suggest though that both store a typeToken exposed through the `FieldRef`'s abstract `getTypeToken()` so that you don't have to cast in your implementation code.

> The reason I make Primitive and Boxed extennd FieldRef is that no matter they are primitive or object, they are field references

sgtm

> CreationContext is also an container Map<FieldRef<?>, Object> ctx which contains all the values
This would be a performance challenge later as we don't want to "auto-box" primitives

> We wouldn't create ObjectDecoderContextIml again, but only update the values each FieldRef pointing to

Let's make sure this approach works with recursively defined types and does not overwrite things in unexpected ways, i.e. the type like the following might have a problem:

```java
class Foo {
  String field;
  @Nullable Foo foo;
```

Reply:
That is a very good call!
Actually it will take care of these recursively defined types well after I enable it to read `null` since each object will have its own `CreationContext` and it won't overwrite things.

Reply:
Also, `TypeToken` didn't support `TypeVariable` right now, but in order to support decode nested Generic type it should. I will have a PR for that first, then come back to support the recursively defined type.